python3Packages.pysmartapp, python3Packages.pysmartthings, python3Packages.smhi, python3Packages.homeconnect

ulrikstrid opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change
I needed a few packages for my home assistant setup.

###### Things done

Added `homeconnect`, `pysmartapp`, `pysmartthings` and `smhi` python packages


This is my first PR to nixpkgs, I don't know what's the best practice for adding multiple packages.
I'm not a Python developer and I'm not sure I have done all the best practice stuff.

Replaces #119555

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [x] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

ulrikstrid committed:
maintainers: add ulrikstrid

ulrikstrid committed:
python3Packages.homeconnect: init at 0.6.3

Co-authored-by: Fabian Affolter <mail@fabian-affolter.ch>

ulrikstrid committed:
python3Packages.pysmartapp: init at 0.3.3

Co-authored-by: Fabian Affolter <mail@fabian-affolter.ch>

ulrikstrid committed:
python3Packages.pysmartthings: init at 0.7.6

Co-authored-by: Fabian Affolter <mail@fabian-affolter.ch>

ulrikstrid committed:
python3Packages.smhi: init at 1.0.13

Co-authored-by: Fabian Affolter <mail@fabian-affolter.ch>

fabaff commented on pkgs/development/python-modules/smhi/default.nix:
```diff
@@ -0,0 +1,55 @@
+  ];
+
+  checkInputs = [
+    black
+    coverage
+    flake8
+    mock
+    pyflakes
+    pylint
```
At first glance, I guess that the linting packages are not used and needed to run the tests.

```suggestion
    coverage
    mock
```

`mock` could probably also be a leftover.

fabaff commented on pkgs/development/python-modules/smhi/default.nix:
```diff
@@ -0,0 +1,55 @@
+  };
+
+  propagatedBuildInputs = [
+    aiohttp
+    requests
+    requests-toolbelt
```
It looks like that the project moved from `requests` to `aiohttp` at some point.

```suggestion
    aiohttp
```

fabaff commented on pkgs/development/python-modules/homeconnect/default.nix:
```diff
@@ -0,0 +1,31 @@
+    requests_oauthlib
+  ];
+
+  pythonImportsCheck = [ "homeconnect" ];
```
```bash
Check whether the following modules can be imported: homeconnect
setuptoolsCheckPhase
Executing setuptoolsCheckPhase
running test
WARNING: Testing via this command is deprecated and will be removed in a future version. Users looking for a generic test entry point independent of test runner are encouraged to use tox.
running egg_info
writing homeconnect.egg-info/PKG-INFO
writing dependency_links to homeconnect.egg-info/dependency_links.txt
writing requirements to homeconnect.egg-info/requires.txt
writing top-level names to homeconnect.egg-info/top_level.txt
reading manifest file 'homeconnect.egg-info/SOURCES.txt'
reading manifest template 'MANIFEST.in'
writing manifest file 'homeconnect.egg-info/SOURCES.txt'
running build_ext

----------------------------------------------------------------------
Ran 0 tests in 0.000s

OK
[...]
```

Thus, 

```suggestion
  # Project has no tests
  doCheck = false;
  pythonImportsCheck = [ "homeconnect" ];
```

r-rmcgibbo commented:
Result of `nixpkgs-review pr 119628` at 37a474c5 run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>4 packages failed to build:</summary>
  <ul>
    <li><a href="https://gist.github.com/3ead7bf6f37e1a2517f1c849165e985b">python38Packages.pysmartapp</a></li>
    <li><a href="https://gist.github.com/5d7b3b2ec2f77c907f2a9cdead61e6f5">python38Packages.pysmartthings</a></li>
    <li><a href="https://gist.github.com/eafbb1be5d06fecb9b19320c131bcadf">python39Packages.pysmartapp</a></li>
    <li><a href="https://gist.github.com/aa51c56b3f6367831b5da30f11b33ad2">python39Packages.pysmartthings</a></li>
  </ul>
</details>
<details>
  <summary>4 packages built successfully:</summary>
  <ul>
    <li>python38Packages.homeconnect</li>
    <li>python38Packages.smhi</li>
    <li>python39Packages.homeconnect</li>
    <li>python39Packages.smhi</li>
  </ul>
</details>
<details>
  <summary>3 suggestions:</summary>
  <ul>
    <li>warning: unused-argument

Unused argument: `six`.
Near pkgs/development/python-modules/pysmartapp/default.nix:4:3:
```
  |
4 | , six
  |   ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/no-python-tests.md">no-python-tests</a>

Test runner could not discover any test cases: ‘Ran 0 tests in 0.000s’
Near pkgs/development/python-modules/homeconnect/default.nix:25:0:
```
   |
25 |     description = "Simple Python client for the BSH Home Connect REST API implementing OAuth 2 authentication, REST calls, and SSE event stream parsing";
   | ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/no-python-tests.md">no-python-tests</a>

Test runner could not discover any test cases: ‘Ran 0 tests in 0.000s’
Near pkgs/development/python-modules/smhi/default.nix:49:0:
```
   |
49 |     description = "Python library for accessing SMHI open forecast data";
   | ^
```
</li>
  </ul>
</details>

Note that build failures [may predate](https://github.com/nix-community/hydra-check) this PR, and could be nondeterministic or hardware dependent.
Please exercise your independent judgement.

fabaff commented on pkgs/development/python-modules/pysmartapp/default.nix:
```diff
@@ -0,0 +1,30 @@
+  pname = "pysmartapp";
+  version = "0.3.3";
+
+  src = fetchPypi {
+    inherit pname version;
+    sha256 = "1i5f6as4f0jnba2zynni0v0y9fkf2qfavpyz73nk78as4j8ysnr9";
+  };
```
```suggestion
  src = fetchFromGitHub {
    owner = "andrewsayre";
    repo = pname;
    rev = version;
    sha256 = "03wk44siqxl15pa46x5vkg4q0mnga34ir7qn897576z2ivbx7awh";
  };
```

fabaff commented on pkgs/development/python-modules/pysmartapp/default.nix:
```diff
@@ -0,0 +1,30 @@
+  propagatedBuildInputs = [
+    httpsig
+  ];
+
+  pythonImportsCheck = [ "pysmartapp" ];
```
Without `doCheck = false;` the build process fails.

```bash
Check whether the following modules can be imported: pysmartapp
setuptoolsCheckPhase
Executing setuptoolsCheckPhase
running test
WARNING: Testing via this command is deprecated and will be removed in a future version. Users looking for a generic test entry point independent of test runner are encouraged to use tox.
WARNING: The directory '/homeless-shelter/.cache/pip' or its parent directory is not owned or is not writable by the current user. The cache has been disabled. Check the permissions and owner of that directory. If executing pip with sudo, you may want sudo's -H flag.
[...]
WARNING: Retrying (Retry(total=0, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.HTTPSConnection object at 0x7ffff5ef5640>: Failed to establish a new connection: [Errno -2] Name or service not known')': /simple/tox/
ERROR: Could not find a version that satisfies the requirement tox<4.0.0,>=3.5.0
ERROR: No matching distribution found for tox<4.0.0,>=3.5.0
error: Command '['/nix/store/d44wd6n98f93hjr6q1d1phhh1hw7a17d-python3-3.8.8/bin/python3.8', '-m', 'pip', '--disable-pip-version-check', 'wheel', '--no-deps', '-w', '/build/tmpwigins6y', '--quiet', 'tox<4.0.0,>=3.5.0']' returned non-zero exit status 1.
builder for '/nix/store/nxy7d2bk1i0q7r0cy5033sdgdxzm6c65-python3.8-pysmartapp-0.3.3.drv' failed with exit code 1
error: build of '/nix/store/nxy7d2bk1i0q7r0cy5033sdgdxzm6c65-python3.8-pysmartapp-0.3.3.drv' failed
```

But the upstream repo contains tests. Switch to the source from GitHub (see above) and add the test requirements as `checkInputs` and as inputs.

```suggestion
  checkInputs = [
    pytest-asyncio
    pytestCheckHook
  ];

  pythonImportsCheck = [ "pysmartapp" ];
```

fabaff commented on pkgs/development/python-modules/pysmartthings/default.nix:
```diff
@@ -0,0 +1,29 @@
+  pname = "pysmartthings";
+  version = "0.7.6";
+
+  src = fetchPypi {
```
Switch to the source from GitHub. Check `test-requirements.txt` for the needed inputs.

```bash
Check whether the following modules can be imported: pysmartthings
setuptoolsCheckPhase
Executing setuptoolsCheckPhase
running test
WARNING: Testing via this command is deprecated and will be removed in a future version. Users looking for a generic test entry point independent of test runner are encouraged to use tox.
WARNING: The directory '/homeless-shelter/.cache/pip' or its parent directory is not owned or is not writable by the current user. The cache has been disabled. Check the permissions and owner of that directory. If executing pip with sudo, you may want sudo's -H flag.
WARNING: Retrying (Retry(total=4, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.HTTPSConnection object at 0x7ffff61dd520>: Failed to establish a new connection: [Errno -2] Name or service not known')': /simple/tox/
WARNING: Retrying (Retry(total=3, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.HTTPSConnection object at 0x7ffff61dd370>: Failed to establish a new connection: [Errno -2] Name or service not known')': /simple/tox/
WARNING: Retrying (Retry(total=2, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.HTTPSConnection object at 0x7ffff61dd4f0>: Failed to establish a new connection: [Errno -2] Name or service not known')': /simple/tox/
WARNING: Retrying (Retry(total=1, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.HTTPSConnection object at 0x7ffff61ddc10>: Failed to establish a new connection: [Errno -2] Name or service not known')': /simple/tox/
WARNING: Retrying (Retry(total=0, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.HTTPSConnection object at 0x7ffff61dda60>: Failed to establish a new connection: [Errno -2] Name or service not known')': /simple/tox/
ERROR: Could not find a version that satisfies the requirement tox<4.0.0,>=3.5.0
ERROR: No matching distribution found for tox<4.0.0,>=3.5.0
[...]
```

fabaff commented on pkgs/development/python-modules/smhi/default.nix:
```diff
@@ -0,0 +1,55 @@
+  pname = "smhi";
+  version = "1.0.13";
+
+  src = fetchPypi {
```
The source tarball published on PyPI doesn't contains the tests but the source from GitHub does. It looks like that it will not be possible to run the tests as they want to communicate with `opendata-download-metfcst.smhi.se:443`.

If that's the case add `doCheck = false;` and remove all test related inputs.

ulrikstrid replied:
Yes, tests communicate with that url

ulrikstrid commented:
@fabaff I have now addressed all your comments, do I rebase/squash it back to 5 commits?

fabaff commented:
You have to split commit d745cea according to the package which is touched and then squash it. In the current situation it would not be possible to revert that commit without additional work if there is a issue with one packages. 

The commits should be prefixed with `python3Packages.`, e.g., `python3Packages.homeconnect: init at 0.6.3`.

> This is my first PR to nixpkgs, I don't know what's the best practice for adding multiple packages.

Usually it's more efficient to stick to one topic/issue/package (incl. dependencies) in a PR but there is no hard requirement to do it that way. Smaller PR are easier to review and to update.

For Home Assistant dependencies it will be required to update the `component-packages` listing and if possible to enable the tests. This means that 8 more commits could be needed here. Sure, collecting multiple packages in one PR save some time when running `nix-review` but there could be a lot of rebasing and wasted time for debugging build failures.

ulrikstrid commented:
I have now rebased and split the commit to the correct commits. Also fixed the names of the commits.

If anyone wants to take these over feel free to open new PRs and set yourself as a maintainer since I'm not a Python developer anyway.

fabaff commented:
You don't have to be a Python developer to maintain Python packages. I maintain some Rust packages but in my life I have probably written 2000 lines of Rust code, thus I'm not sure if that even puts me in the "beginner group".  

We are almost there. The heavy lifting is done. What's left is partially the part for Home Assistant.

For Home Assistant dependencies it's a three steps process: 

- Add Python packages
- Run `./pkgs/servers/home-assistant/parse-requirements.py` to get the dependency tracking (`component-packages`) updated
- Enable Home Assistant's tests for the integration if possible. 

As step 2 and 3 are not related to the Python module itself we put them in separate commits. This gets us three commits to keep the changes small and the option to fast revert if we need to. E.g., Home Assistant dependencies may be removed but this doesn't mean that we want to remove the package itself. Or the tests start to fail.

ulrikstrid commented:
Hello, I don't have any time to fix this properly in the near future it seems. What is the minimal thing I can do to get this merged? Just revert the home-assistant stuff?

SuperSandro2000 commented on pkgs/servers/home-assistant/component-packages.nix:
```diff
@@ -351,7 +351,7 @@
     "hive" = ps: with ps; [ ]; # missing inputs: pyhiveapi
     "hlk_sw16" = ps: with ps; [ ]; # missing inputs: hlk-sw16
-    "home_connect" = ps: with ps; [ aiohttp-cors ]; # missing inputs: homeconnect
+    "home_connect" = ps: with ps; [ aiohttp-cors homeconnect ];
```
Can we enable tests for any of them?

fabaff replied:
The `home_connect` has tests. It's worth a try, especially as they are already present for a while.

fabaff commented:
Let me cherry-pick the commits and do the tasks needed for Home Assistant.

It's kind of my fault that we ended here. My mentoring was bad. Sorry.

mweinelt commented:
I think it was a bad idea to do so many packages at once. It makes mentoring very hard, given there are so many things to critique for first time contributors.

fabaff commented:
- homeconnect: https://github.com/NixOS/nixpkgs/pull/120532

fabaff commented:
@ulrikstrid, would you like to maintain the packages after they are included?

ulrikstrid commented:
> @ulrikstrid, would you like to maintain the packages after they are included?

I don't have strong feelings either way. I'll probably try to update them if my setup breaks. But anyone can feel free to take it over.

Thanks for taking over the PRs ❤️

fabaff commented:
This was done. Thanks @ulrikstrid for your contributions.
